fix: Resolve security issue with blacken

flodet-me opened the pull request:
Currently, blacken function replaces the characters with the replacement without considering that the length can already be an issue, as this could expose sensible information or at least can result in that the malicious actor can gain further information about the data.

Resolve this issue by adding a default length for blacken that can resolve most cases. If intended blacken = 5 this would not really help, but anything else are even less useful.

I considered Random and fallback to default length based on previous data.

Due to the lack of knowledge of previous data, we can not use previous data to always use X length without rework.

Random uses more computing power and is not really adding more security.

heutelbeck commented:
Request breaks tests. Please fix. Please ensure the original behaviour is also still an option, as it is required for some use-cases. 
Thank you.

flodet-me commented:
I've adjusted the implementation based on the Feedback.

I'm now using a parameter for providing a custom length that should the blacken characters be set. If it is not set, it should just replace the to replace / blacken characters. This is currently implemented a bit hacky, as I can't create a nullable primitive, which I would use in my main language c# and I would like not to create an artificial wrapper class just for this.

I also considered and implemented the usage of an enum to define different types of  blacken in the invocation, but it felt like it would be a big burden for developers to correctly apply and currently more options are not really required.

I further adjusted code, that I found not really intuitive and not optimally written.

TODO:

- Write more and accurate unit test that validate the blacken implementation in the different scenarios
- Consultation of you if this is a valid implementation.

If this kind of function (blacken...) should be used in many forms it might also be a good idea to provide a capability to automatically transform and validate the parameters. This could be either done by a library or custom made.

What do you think?

StefWe commented:
Do the adjustments to blacken() need to be considered [here](https://github.com/heutelbeck/sapl-policy-engine/blob/master/sapl-spring-security/src/main/java/io/sapl/spring/constraints/providers/ContentFilterUtil.java#L421) as well?
